// ng2-tag-input/dist/modules/components/tag/tag-ripple.component.js
import { Component, animate, trigger, style, transition, keyframes, state, Input } from '@angular/core';
export var TagRipple = (function () {
    function TagRipple() {
        this.state = 'none';
    }
    TagRipple.decorators = [
        { type: Component, args: [{
                    selector: 'tag-ripple',
                    styles: ["\n        :host {\n            width: 100%;\n            height: 100%;\n            left: 0;\n            overflow: hidden;\n            position: absolute;\n        }\n        \n        .tag-ripple {\n            background: rgba(0, 0, 0, 0.1);\n            top: 50%;\n            left: 50%;\n            height: 100%;\n            transform: translate(-50%, -50%);\n            position: absolute;\n        }\n    "],
                    template: "\n        <div class=\"tag-ripple\" [@ink]=\"state\"></div>\n    ",
                    animations: [
                        trigger('ink', [
                            state('none', style({ width: 0, opacity: 0 })),
                            transition('none => clicked', [
                                animate(200, keyframes([
                                    style({ opacity: 1, offset: 0, width: 0, height: 0, borderRadius: '100%' }),
                                    style({ opacity: 1, offset: 0.15, width: '15%', height: '30%' }),
                                    style({ opacity: 1, offset: 0.35, width: '35%', height: '50%' }),
                                    style({ opacity: 1, offset: 0.5, width: '50%', height: '75%' }),
                                    style({ opacity: 1, offset: 0.8, width: '80%', height: '90%' }),
                                    style({ opacity: 1, offset: 1, width: '100%', height: '100%', borderRadius: '16px' })
                                ]))
                            ])
                        ])
                    ]
                },] },
    ];
    TagRipple.ctorParameters = function () { return []; };
    TagRipple.propDecorators = {
        'state': [{ type: Input },],
    };
    return TagRipple;
}());
